binding.cc

```diff
@@ -565,24 +574,53 @@ struct Iterator {
     }
   }
 
+  /**
+   * Maintain a reference counter to prevent garbage collection of the iterator
+   * while nexting or while we're closing the database. Keep track of non-gc'ed
+   * iterators to explicitly close them before closing the database.
+   */
   void Attach (napi_ref ref) {
     ref_ = ref;
     database_->AttachIterator(id_, this);
   }
 
-  napi_ref Detach () {
-    database_->DetachIterator(id_);
-    return ref_;
+  /**
+   * This can be called by the destructor (triggered by JS garbage collection)
+   * and/or by the database's CloseWorker (on db close). If first called by the
+   * destructor that will be the only call, because afterwards the database no
+   * longer knows of the iterator. If first called by the CloseWorker, skip the
+   * subsequent destructor's call. Note that we don't release all associated
+   * resources here; only those that would prohibit closing the database. The
+   * destructor always does the last and final cleanup.
+   */
+  void Close () {
+    if (!closed_) {
+      closed_ = true;
+      delete dbIterator_;
+      dbIterator_ = NULL;
+      database_->ReleaseSnapshot(options_->snapshot);
+      database_->DetachIterator(id_);
+    }
   }
 
-  leveldb::Status IteratorStatus () {
-    return dbIterator_->status();
+  void DeleteReference (napi_env env) {
+    // TODO: we seem to be hitting nodejs/node#24494 or a related bug, in node < 11.
+    // When uncommented we get a segfault in the finalizer caller.
+    // napi_delete_reference(env, ref_);
```

Comment:
I don't know how to fix this. We need to delete the reference to avoid leaking, but calling `napi_delete_reference` in the finalizer callback results in a segfault (in node 8 and 10). Might be https://github.com/nodejs/node/pull/24494?